### PR TITLE
CON-2185-bring api sandbox to function app.

### PR DESCRIPTION
**Pull request by shomavg**

### CODE REVIEW

File: src/SFA.DAS.Assessor.Functions.Infrastructure/ConfigurationHelper.cs [hunk 1]
```diff
@@ -13,5 +20,35 @@ public static string GetAppName(IConfiguration configuration)
         {
             return configuration.GetConnectionStringOrSetting("AppName");
         }
+
+        public static IWebConfiguration GetConfiguration()
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions.Infrastructure/ConfigurationHelper.cs:24]: What is this???

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]
You can't just copy this across from one place to another you need to make it fit with how the new function library works for configuration, logging function names, trigger time configuration etc...

File: src/SFA.DAS.Assessor.Functions/Startup.cs [hunk 2]
```diff
@@ -55,7 +57,7 @@ public override void Configure(IFunctionsHostBuilder builder)
             builder.Services.Configure<AssessorApiAuthentication>(config.GetSection("AssessorApiAuthentication"));            
             builder.Services.Configure<CertificateDetails>(config.GetSection("CertificateDetails"));
             builder.Services.Configure<SftpSettings>(config.GetSection("SftpSettings"));
-
+            builder.Services.Configure<ExternalApiDataSyncConfig>(config.GetSection("ExternalApiDataSyncConfig"));
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions/Startup.cs:60]: Remove Config word from here if you can

File: src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/ExternalApiDataSyncFunction.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        private readonly IExternalApiDataSyncCommand _command;
+
+        public ExternalApiDataSyncFunction(IExternalApiDataSyncCommand command)
+        {
+            _command = command;
+        }
+
+        [FunctionName("ExternalApiDataSyncFunction")]
+        public async Task Run([TimerTrigger("%ExternalApiDataSyncFunctionSchedule%", RunOnStartup = true)]TimerInfo myTimer, ILogger log, ExecutionContext context)
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/ExternalApiDataSyncFunction.cs:19]: This "ExternalApiDataSyncFunctionSchedule" is defined in another file with an actual schedule - which you have from the original code - you can't see that when the project is load - swap to directory view

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]
Just a few minor tweaks needed

File: src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/AssessorServiceApiClient.cs [hunk 4]
```diff
@@ -183,6 +183,14 @@ public async Task SendEmailWithTemplate(SendEmailRequest sendEmailRequest)
                 await PostPutRequest(request, sendEmailRequest);
             }
         }
+
+        public async Task ExternalApiDataSync(GetDataSyncRequest getDataSyncRequest)
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/AssessorServiceApiClient.cs:187]: RebuildExternalApiSandbox?

File: src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/ExternalApiDataSyncFunction.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
+using System;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Extensions.Logging;
+using System.Threading.Tasks;
+using SFA.DAS.Assessor.Functions.ExternalApis.Assessor;
+using SFA.DAS.Assessor.Functions.ExternalApis.Assessor.Types;
+
+namespace SFA.DAS.Assessor.Functions.Functions.ExternalApiDataSync
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/ExternalApiDataSyncFunction.cs:8]: RebuildExternalApiSandboxFunction?? - seems like a better name - and you are free to rename it here - just make everything else follow along

File: src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/ExternalApiDataSyncFunction.cs (hunk 3 above)
Comment (ChrisJWoodcock) [reply src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/ExternalApiDataSyncFunction.cs:20]: Don't forget the other files for this schedule need to be amended to include the default schedule

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/ExternalApiDataSyncFunction.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            try
+            {
+                if (myTimer.IsPastDue)
+                {
+                    log.LogInformation("Epao ExternalApiDataSyncFunction timer trigger is running later than scheduled");
+                }
+
+                log.LogInformation($"Epao ExternalApiDataSyncFunction started");
+
+                await _assessorServiceApi.ExternalApiDataSync(dataSyncRequest);
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/ExternalApiDataSyncFunction.cs:32]: use the standard command pattern here even when it does almost nothing - for consistency

File: src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/AssessorServiceApiClient.cs [hunk 7]
```diff
@@ -183,6 +183,14 @@ public async Task SendEmailWithTemplate(SendEmailRequest sendEmailRequest)
                 await PostPutRequest(request, sendEmailRequest);
             }
         }
+
+        public async Task RebuildExternalApiSandbox(RebuildExternalApiSandboxRequest rebuildExternalApiSandboxRequest)
+        {
+            using (var request = new HttpRequestMessage(HttpMethod.Post, $"/api/v1/externalapidatasync"))
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/AssessorServiceApiClient.cs:189]: You can change the url as well to make it clearer

File: src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/RebuildExternalApiSandboxFunction.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        private readonly IRebuildExternalApiSandboxCommand _command;
+
+        public RebuildExternalApiSandboxFunction(IRebuildExternalApiSandboxCommand command)
+        {
+            _command = command;
+        }
+        
+        [FunctionName("RebuildExternalApiSandboxFunction")]
+        public async Task Run([TimerTrigger("%RebuildExternalApiSandboxFunctionSchedule%", RunOnStartup = true)]TimerInfo myTimer, ILogger log, ExecutionContext context)
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions/Functions/ExternalApiDataSync/RebuildExternalApiSandboxFunction.cs:19]: Should not have RunOnStartup for this

File: azure/template.json [hunk 9]
```diff
@@ -1,262 +1,268 @@
...
+      "type": "string",
+      "metadata": {
+        "description": "certificateDeliveryNotificationFunction CRON Schedule e.g. 0 */15 * * * *"
+      },
+      "defaultValue": "0 15 * * *"
+    },
+    "RebuildExternalApiSandboxFunctionSchedule": {
+      "type": "string",
+      "metadata": {
+        "description": "RebuildExternalApiSandboxFunction CRON Schedule e.g. 0 */15 * * * *"
```
Comment (ChrisJWoodcock) [inline azure/template.json:74]: e.g. should be the default I think

File: src/SFA.DAS.Assessor.Functions/Startup.cs [hunk 10]
```diff
@@ -80,7 +79,6 @@ public override void Configure(IFunctionsHostBuilder builder)
             builder.Services.AddTransient<IPrintProcessCommand, PrintProcessCommand>();
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions/Startup.cs:79]: I think your new command should be here?

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

**Review by ChrisJWoodcock — APPROVED** [review]

**Review by cofaulco — APPROVED** [review]

**Pull request merged**